### PR TITLE
0.7.0 release

### PR DESCRIPTION
**Pull request by martyn**
Currently a wip

### CODE REVIEW

File: hypergan/gan.py [hunk 1]
```diff
@@ -1,317 +1,340 @@
...
+
+                
+
+    def init_session(self, device):
+        # Initialize tensorflow
+        with tf.device(device):
+            self.sess = tf.Session(config=tf.ConfigProto())
+
+    def run(self):
+        args = self.args
```
Comment (martyn) [inline hypergan/gan.py]: This object should not know about `args` (let cli handle it).

File: hypergan/gan.py [hunk 2]
```diff
@@ -1,317 +1,340 @@
...
+
+        if args.config is None:
+            filename = '~/.hypergan/configs/'+config['uuid']+'.json'
+            print("[hypergan] saving network configuration to: " + filename)
+            config = self.selector.load_or_create_config(filename, config)
+        else:
+            save_file = "~/.hypergan/saves/"+args.config+".ckpt"
+            config['uuid'] = args.config
+
+        save_file = "~/.hypergan/saves/"+config["uuid"]+".ckpt"
```
Comment (martyn) [inline hypergan/gan.py]: Suggestion: Remove this uuid stuff - that should come from the cli I think.  Lets require networks to be named.

File: hypergan/gan.py [hunk 3]
```diff
@@ -1,317 +1,340 @@
...
+        if(args.method == 'build' or args.method == 'serve'):
+            graph_type = 'generator'
+        else:
+            graph_type = 'full'
+
+        graph = self.create_graph(x, y, f, graph_type, args.device)
+
+        print( "Save file", save_file,"\n")
+        #TODO refactor save/load system
+        if args.method == 'serve':
```
Comment (martyn) [inline hypergan/gan.py]: This shouldn't be here.  'serve' should in a different file.

File: hypergan/gan.py [hunk 4]
```diff
@@ -1,317 +1,340 @@
...
+                print("Model loaded")
+            else:
+                print("No checkpoint file found")
+        else:
+            print(" |= Initializing new network")
+            with tf.device(args.device):
+                init = tf.initialize_all_variables()
+                self.sess.run(init)
+
+        self.output_graph_size()
```
Comment (martyn) [inline hypergan/gan.py]: move to cli

File: hypergan/gan.py [hunk 5]
```diff
@@ -1,317 +1,340 @@
...
+        tf.train.start_queue_runners(sess=self.sess)
```
Comment (martyn) [inline hypergan/gan.py]: Don't start these here - let the input system do it.

File: hypergan/gan.py [hunk 6]
```diff
@@ -1,317 +1,340 @@
...
+        testx = self.sess.run(x)
+
+        if args.method == 'build':
```
Comment (martyn) [inline hypergan/gan.py]: All this should go in the cli too.

File: hypergan/gan.py [hunk 7]
```diff
@@ -1,317 +1,340 @@
...
+                args.directory,
+                args.device,
+                seconds=None,
+                bitrate=None,
+                width=width,
+                height=height,
+                channels=channels,
+                crop=crop
+        )
+        self.config['y_dims']=num_labels
```
Comment (mikkel) [inline hypergan/gan.py]: Overwriting the config object seems wrong here.  The CLI should do it if we need to.

File: hypergan/graph/graph.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        batch_size = config["batch_size"]
+        z_dim = int(config['generator.z'])
+        batch_norm = config['generator.regularizers.layer']
+
+        g_losses = []
+        extra_g_loss = []
+        d_losses = []
+
+        #initialize with random categories
+        categories = [random_category(config['batch_size'], size, config['dtype']) for size in config['categories']]
```
Comment (mikkel) [inline hypergan/graph/graph.py]: Could this be seen as a separate z encoder?

Like
```bash
  "z_encoders": [linear.config(), gaussian.config(), categories.config(5)]
```

File: hypergan/graph/graph.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        if(len(categories) > 0):
+            categories_t = [tf.concat(1, categories)]
+        else:
+            categories_t = []
+
+        z, encoded_z, z_mu, z_sigma = config['encoder'](config, x, y)
+
+        # create generator
```
Comment (mikkel) [inline hypergan/graph/graph.py]: remove comment

File: hypergan/graph/graph.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        g = self.generator([y, z]+categories_t)
+
+        #encoded = generator([y, encoded_z]+categories_t, reuse=True)
```
Comment (mikkel) [inline hypergan/graph/graph.py]: do we want to support autoencoding in this version or just cut it out of the base?  Would it make sense as an api example?

File: hypergan/graph/graph.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+                                           - tf.exp(z_sigma), 1)
+
+            latent_loss = tf.reshape(latent_loss, [int(latent_loss.get_shape()[0]), 1])
+        else:
+            latent_loss = None
+        np_fake = np.array([0]*config['y_dims']+[1])
+        fake_symbol = tf.tile(tf.constant(np_fake, dtype=config['dtype']), [config['batch_size']])
+        fake_symbol = tf.reshape(fake_symbol, [config['batch_size'],config['y_dims']+1])
+
+        #real_symbols = tf.concat(1, [y, tf.zeros([config['batch_size'], 1])])
```
Comment (mikkel) [inline hypergan/graph/graph.py]: remove?

File: hypergan/graph/graph.py [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+
+
+        zeros = tf.zeros_like(d_fake_sig, dtype=config['dtype'])
+        ones = tf.zeros_like(d_real_sig, dtype=config['dtype'])
+
+        generator_target_prob = config['g_target_prob']
+        d_label_smooth = config['d_label_smooth']
+        d_fake_loss = tf.nn.sigmoid_cross_entropy_with_logits(d_fake_sig, zeros)
+        d_real_loss = sigmoid_kl_with_logits(d_real_sig, 1.-d_label_smooth)
+        #if(config['adv_loss']):
```
Comment (mikkel) [inline hypergan/graph/graph.py]: remove?

File: hypergan/graph/graph.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+        else:
+            latent_loss = None
+        np_fake = np.array([0]*config['y_dims']+[1])
+        fake_symbol = tf.tile(tf.constant(np_fake, dtype=config['dtype']), [config['batch_size']])
+        fake_symbol = tf.reshape(fake_symbol, [config['batch_size'],config['y_dims']+1])
+
+        #real_symbols = tf.concat(1, [y, tf.zeros([config['batch_size'], 1])])
+        real_symbols = y
+
+
```
Comment (mikkel) [inline hypergan/graph/graph.py]: Loss functions should be composable.

```
"losses": [ generator.loss.config(), discriminator.loss.config(), custom.loss.config() ]
```
